internal/pkg/agent/application/upgrade/marker_watcher.go

```diff
@@ -0,0 +1,124 @@
+// Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+// or more contributor license agreements. Licensed under the Elastic License;
+// you may not use this file except in compliance with the Elastic License.
+
+package upgrade
+
+import (
+	"context"
+	"fmt"
+	"path/filepath"
+
+	"github.com/elastic/elastic-agent/pkg/core/logger"
+
+	"github.com/fsnotify/fsnotify"
+)
+
+type MarkerWatcher interface {
+	Watch() <-chan UpdateMarker
+	Errors() <-chan error
+	Run(ctx context.Context) error
+	Close() error
+}
+
+type MarkerFileWatcher struct {
+	watcher        *fsnotify.Watcher
+	markerFilePath string
+
+	logger *logger.Logger
+
+	updateCh chan UpdateMarker
+	errCh    chan error
+}
+
+func newMarkerFileWatcher(upgradeMarkerFilePath string, logger *logger.Logger) (MarkerWatcher, error) {
+	watcher, err := fsnotify.NewWatcher()
+	if err != nil {
+		return nil, fmt.Errorf("failed to create upgrade marker watcher: %w", err)
+	} // Watch the upgrade marker file's directory, not the file itself, so we
+
+	// notice the file even if it's deleted and recreated.
```

Comment:
[Suggestion]
I guess those are the same comment, right?
```suggestion
	} 
	
	// Watch the upgrade marker file's directory, not the file itself, so we
	// notice the file even if it's deleted and recreated.
```

Reply:
Yeah, thanks, not sure what happened here 😄.  Maybe something during a rebase + conflict resolution 🤷.

Reply:
Fixed in ea5177522c58e011b59c354ff55954ef5d69e1eb.